Introduce SchedulingStateMachine for unified scheduler [NO-MERGE&REVIEW-ONLY-MODE]

ryoqun opened the pull request:
#### Problem

The unified scheduler is still single-threaded, because the most important piece of code is still missing: the scheduling code itself.

#### Summary of Changes

Implement it with cleanest way possible and most documented way possible. FINALLY, TIME HAS COME.

Note that there's more work to be done for general availability of unified scheduler after this pr, but it is arguably the most important PR.

#### numbers

##### steps <a id="steps"></a>

1. Download `replaying-bench-mainnet-beta-2024-02.tar.zst`: https://drive.google.com/file/d/1Jc1cd3pHKkaaT1yJrGeBAjBMJpkKS0hc/view?usp=sharing
2. untar
3. run `solana-ledger-tool --ledger ./path/to/replaying-bench-mainnet-beta-2024-02 verify ...`


##### (A) `--block-verification-method blockstore-processor` with this pr (reference):

```
[2024-02-22T15:24:06.957483120Z INFO  solana_ledger::blockstore_processor] ledger processed in 28 seconds, 198 ms, 761 µs and 24 ns. root slot is 249712363, 36 banks: 249712363, 249712364, 249712365, 249712366, 249712367, 249712368, 249712369, 249712370, 249712371, 249712372, 249712373, 249712374, 249712375, 249712378, 249712379, 249712380, 249712381, 249712382, 249712383, 249712384, 249712385, 249712386, 249712387, 249712388, 249712389, 249712390, 249712391, 249712392, 249712393, 249712394, 249712395, 249712396, 249712397, 249712398, 249712399, 249712400
```

##### (B) `--block-verification-method unified-scheduler` with this pr:

**~1.3x** faster than (A):

```
[2024-02-22T15:22:46.686416459Z INFO  solana_ledger::blockstore_processor] ledger processed in 20 seconds, 621 ms, 124 µs and 512 ns. root slot is 249712363, 36 banks: 249712363, 249712364, 249712365, 249712366, 249712367, 249712368, 249712369, 249712370, 249712371, 249712372, 249712373, 249712374, 249712375, 249712378, 249712379, 249712380, 249712381, 249712382, 249712383, 249712384, 249712385, 249712386, 249712387, 249712388, 249712389, 249712390, 249712391, 249712392, 249712393, 249712394, 249712395, 249712396, 249712397, 249712398, 249712399, 249712400
```

##### (C) (fyi) #33070 (all extra opt is applied):

**~1.8x** faster than (A), **~1.3x** faster than (B):

```
[2024-02-22T15:16:51.304579110Z INFO  solana_ledger::blockstore_processor] ledger processed in 15 seconds, 399 ms, 459 µs and 572 ns. root slot is 249712363, 36 banks: 249712363, 249712364, 249712365, 249712366, 249712367, 249712368, 249712369, 249712370, 249712371, 249712372, 249712373, 249712374, 249712375, 249712378, 249712379, 249712380, 249712381, 249712382, 249712383, 249712384, 249712385, 249712386, 249712387, 249712388, 249712389, 249712390, 249712391, 249712392, 249712393, 249712394, 249712395, 249712396, 249712397, 249712398, 249712399, 249712400
```


#### context

extracted from #33070

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -1023,7 +1062,7 @@ mod tests {
             Ok(_)
         );
-        scheduler.schedule_execution(&(good_tx_after_bad_tx, 0));
+        scheduler.schedule_execution(&(good_tx_after_bad_tx, 1));
```
this is due to task_index is started to be `assert!()`-ed....

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+    }
+
+    #[must_use]
+    pub fn schedule_task(&mut self, task: Task) -> Option<Task> {
```
Additionally it's unclear to me why we want to return the task at all here instead of just returning something a little bit more direct such as a bool indicating it can be immediately sent as a payload. since we do not modify the task in case of success, that seems fine?

Maybe you have plans to do some modifications to Task in here in the future?

ryoqun replied:
> Maybe you have plans to do some modifications to Task in here in the future?

yeah. I'm planning to reuse `Arc::strong_count()` instead of `blocked_task_count()`. I'll update the docs.

apfitzge reviewed (commented):
Did an initial scan through the code, appreciate the well thought out comments! I will have to give this another pass, as well as getting through the tests.

ryoqun commented:
@apfitzge thanks for reviewing. I'll later address them.

I've added benchmark steps in the pr description.

also, I'll share some on-cpu flame-graph by `perf` to make I'm not crazy. The charts was from different `solana-ledger-tool` invocations timed manually. so, the _proportion_ of the spend walltime is what it's comparable.

#### the scheduler thread of rebase pr

![image](https://github.com/solana-labs/solana/assets/117807/dd935d92-a43f-494b-8aed-268947d3f92a)

#### the schedule thread of this pr

![image](https://github.com/solana-labs/solana/assets/117807/f874a47c-0202-4cba-a326-3ec49a4db3b7)


#### some random remarks

- `SchedulingStateMachine::{de,}schedule_task()` and crossbeam select is what it's absolutely worth to burn cycles.
- the rebase pr doesn't have any extra overhead, very close to the ideal runtime profile of the scheduler thread
- this pr has bunch of _overhead_ by dropping the task in thread and accumulating the timings. (this is why i want to introduce the accumulator thread)

ryoqun commented:
also, off-cpu flame graph shows there's no idling other then the expected crossbeam select:

![image](https://github.com/solana-labs/solana/assets/117807/46604724-0a9d-4e83-812e-3e1881e5563c)

as a quick refresher of this unusual off-cpu flame graph, you can see some lock-contension bolow (there's known one in solana-runtime, which is fixed at the rebase pr, but not yet in this pr):

![image](https://github.com/solana-labs/solana/assets/117807/58981247-d11b-475d-971a-23ab5e287cb8)

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+    pub(super) struct Token<V: 'static>(PhantomData<*mut V>);
+
+    impl<V> Token<V> {
+        // Returns the token to acquire a mutable reference to the inner value of [TokenCell].
+        //
+        // Safety:
+        // This method should be called exactly once for each thread at most.
```
hm? This pr isn't using `RefCell`. it's using `UnsafeCell`. `RefCell` has runtime check, which i'd like to avoid.

apfitzge replied:
In the `thread_local!`, `TOKENS` is a `RefCell<BTreeSet<_>>`:
```rust
thread_local! {
    static TOKENS: RefCell<BTreeSet<TypeId>> = const { RefCell::new(BTreeSet::new()) };
}
```
just wanted to clarify whether this fn `assume_exclusive_mutating_thread` cannot panic on the mutable borrow of the RefCell since it cannot be called by same thread multiple times at same time. So the unsafeness is coming from the additional constraint that the caller must not call this more than once per thread for the same type.

ryoqun replied:
thanks for clarifying your review comment. I fixed the doc comments: 087e499c4962afc42ff8315ce102ee1add2a842a

> just wanted to clarify whether this fn `assume_exclusive_mutating_thread` cannot panic on the mutable borrow of the RefCell since it cannot be called by same thread multiple times at same time.

yeah. this is correct.



> So the unsafeness is coming from the additional constraint that the caller must not call this more than once per thread for the same type.

yeah. this is correct as well. I clarified the doc comment.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+//! tasks while deschduling, in addition to the above-mentioned book-keeping processing. Namely,
+//! when given address is ready for new fresh locking resulted from descheduling a task (i.e. write
+//! lock is released or read lock count is reached to zero), it pops out the first element of the
+//! FIFO blocked-task queue of the address. Then, it immediately marks the address as relocked. It
```
>  it seems that only the first read in the queue is unblocked, but I'm hoping that is not the case of the actual implementation.

yeah, impl aggressively unblock all consecutive readable tasks. will clarity the doc part.

ryoqun replied:
> will clarify the doc part.

done: 4bdd8bd04e91f456f6811cef16b68fad57a1a8a2

apfitzge commented:
I think I finally have a reasonably good understanding of how your scheduling works (much easier with trimmed down PR! ❤️ ).
As I'm thinking on this, I think this state-machine and the prio-graph I've used in the new scheduler for block-production are similar in many ways, although there are a few differences. Just gonna jot some thoughts here.

## Similarities

- Use FIFO fed order (in replay case this is record order, for production this is priority-order from a queue)
- Keep track of access-kind per account in "use" 
- Pop unblocked txs

## Differences

- PG essentially creates a linked-list of transaction dependencies to only **next** transaction.
- US keeps track of current access, and **all** blocked transactions/access type as an ordered list **per transaction**
- Blocked tx resolve time: PG resolves at insert time, US resolves at deschedule time.
- "use" differs since the US is used in a stateful manner, while PG is created on-demand
- PG has a mechanism for allowing user-defined ordering for "unblocked" txs


### Example

Think visual examples always serve better than words here. Let's take a really simple example with 3 conflicting write transactions.


Prio-graph will effectively store a linked list, edges from 1 to 2 to 3.
```mermaid
graph LR;
Tx1 --> Tx2 --> Tx3
```

US will initially store access of Tx1, and list w/ order (Tx2, Tx3). Only when Tx1 is descheduled will the conflict between 2 and 3 be "realized".
```mermaid
graph LR;
Tx1 --> Tx2["Tx2(0, w)"] & Tx3["Tx3(1, w)"];
```

### Closing

Not sure anything I commented here has any actual value, just wanted to share my thoughts that these 2 approaches are remarkably similar but different in their implmentation details due to different uses (stateful vs lazy).

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+/// which [`Task`]s are blocked to be executed after the current task is notified to be finished
+/// via [`::deschedule_task`](`SchedulingStateMachine::deschedule_task`)
+#[derive(Debug)]
+struct PageInner {
```
is there additional data that will be stored on these pages? I think this naming confused me previously. Wonder if something like:
- `AccessList`
- or `UsageList`
is more descriptive?

ryoqun replied:
@apfitzge (once we settled on this, I'll do the once for all renames; so pinging).

apfitzge replied:
Even with ping I somehow didn't get a github alert - sorry!

out of that list I think I prefer `UsageQueue` the most, or maybe even `LockQueue` (slightly shorter). Not as short as `Page` or `Lock` but I think the fact we use it as  a queue is an important description to be in the name.

ryoqun replied:
thanks for the update. kk, let's go with `UsageQueue`. :)

ryoqun replied:
this is done: 725e314efa65357ba9e670f298eacd16d4888b4e and 50b761652d268d837eb1f8b288a57c9f4dbd107e

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
         &self.transaction
     }
+
+    fn lock_attempts(&self) -> &Vec<LockAttempt> {
```
nope at all: 30542c8e951af8a2571e93d58c7fbd33c70fad65

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+        /// complex handling of non-`'static` heaped data in general. Instead, it's manually
+        /// required to ensure this instance is accessed only via its associated Token for the
+        /// entire lifetime.
+        // non-const to forbid unprotected sharing via static variables among threads.
```
done: a6164a90b2e49c80165c5fb62c9379f50e8e249f

ryoqun commented:
> I think I finally have a reasonably good understanding of how your scheduling works (much easier with trimmed down PR! ❤️ ).
> As I'm thinking on this, I think this state-machine and the prio-graph I've used in the new scheduler for block-production are similar in many ways, although there are a few differences. Just gonna jot some thoughts here. 
> ....

thanks for great comparison write-up. I'll check out prio-graph impl later with those good guides in mind. :)

> I am somewhat curious to try plugging in prio-graph to your benches to see how it compares for exactly the same benches you have. Do you have any benches on _just_ the scheduler logic?

yeah: https://github.com/solana-labs/solana/pull/33070#discussion_r1502769937

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+            RequestedUsage::Writable => PageUsage::Writable,
+        }
+    }
+}
```
done: 1d99c444aa18708cfe3f8c52a12fb15744a4d2ef

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1333 @@
+    fn default() -> Self {
+        Self {
+            usage: PageUsage::default(),
+            blocked_tasks: VecDeque::with_capacity(1024),
```
good point. i temporarily reduced it and document about why: 5887a91d5153b7f6a929d9e68c225ec63528ea3a

ryoqun commented on runtime/src/bank.rs:
```diff
@@ -4235,6 +4235,9 @@ impl Bank {
         transaction: &'a SanitizedTransaction,
     ) -> TransactionBatch<'_, '_> {
         let tx_account_lock_limit = self.get_transaction_account_lock_limit();
+        // Note that switching this to .get_account_locks_unchecked() is unacceptable currently.
+        // The unified scheduler relies on the checks enforced here.
+        // See a comment in SchedulingStateMachine::create_task().
```
as this promise is brittle, maybe I'll add test at unified-scheduler-pool.

ryoqun replied:
done: 24cff794fad41bc639de9dd789064a65d195cb2f

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1387 @@
+
+const_assert_eq!(mem::size_of::<TokenCell<PageInner>>(), 40);
+
+/// Scheduler's internal data for each address ([`Pubkey`](`solana_sdk::pubkey::Pubkey`)). Very
+/// opaque wrapper type; no methods just with [`::clone()`](Clone::clone) and
+/// [`::default()`](Default::default).
+#[derive(Debug, Clone, Default)]
+pub struct Page(Arc<TokenCell<PageInner>>);
+const_assert_eq!(mem::size_of::<Page>(), 8);
```
Arc has a runtime cost for cloning so I'm actually a bit surprised to see it used here for such a critical item - though perhaps it allows a simpler implementation. Do you know how much of your insertion time is spent doing Arc-clone? I imagine its' small relative to the Pubkey hashing?

Still, I wonder if you have thought about or have maybe even tested using an indexable storage for the inner pages, and then using index-references as the stored kind on the tasks, to avoid Arc-cloning.

To be clear I'm saying some sort of set up like:
```rust
pubkey_to_index: DashMap<Pubkey, usize>, // not sure if dashmap or locked hashmap? Just some translation that maps Pubkey -> index.
inner_pages: Vec<Option<InnerPage>>, // use a fixed-capacity vector (or slice) so we never add more pages past some pre-determined limit 
```

apfitzge replied:
I don't want to block on this, just sharing my thought on perf here. It's almost certainly a micro-optimization; but in this sensitive code, any small bit can help!

ryoqun replied:
> 


> Arc has a runtime cost for cloning so I'm actually a bit surprised to see it used here for such a critical item - though perhaps it allows a simpler implementation.

Yeah, I used `Arc` here for simplicity and to ship this pr into master asap.

> Do you know how much of your insertion time is spent doing Arc-clone? 

not directly. but, I can indirectly guess it'll take 2-3ns per `Arc::clone()` when the instance isn't contended from an unrelated bench.

> I imagine its' small relative to the Pubkey hashing?

I haven't tested. but I presume yes. Even more, I'm planning to switch to `Rc` from `Arc`. in that case, it's definitely smaller than the hashing.

As I indirectly indicated earlier (https://github.com/solana-labs/solana/pull/35286#issuecomment-1964383019 and https://github.com/solana-labs/solana/pull/35286#discussion_r1502766967), that optimization will allow us to trim down to merely a single `inc` (without the `lock` prefix) in x86-64 asm terms. i.e. possibly auto-vectorizable plain old 1 ALU instruction. And, mem access is 1 pop with no branching. And there's no work to need to be done for _insertion into the scheduling state machine_ as it's not relevant. If you're referring to the insertions happening with the batched cloning in the case of being blocked in `try_lock_for_task()`, index-references or `Rc` will need 1 mem copy of 1 word-size for each address no matter what.

> Still, I wonder if you have thought about or have maybe even tested using an indexable storage for the inner pages, and then using index-references as the stored kind on the tasks, to avoid Arc-cloning.
> 
> To be clear I'm saying some sort of set up like:
> 
> ```
> ...
> ```

Yeah, i've thought about it. But, I concluded this pr's approach should be faster with gut sense. the index approach introduces another pointer dereference indirection, which i'd like to avoid. However, not benchmarked it.

> I don't want to block on this, just sharing my thought on perf here. It's almost certainly a micro-optimization; but in this sensitive code, any small bit can help!

Thanks for sharing your thoughts. happy to share mine. I hope I've squeezed as much as possible.

apfitzge reviewed (commented):
Another short round of comments - I still need to take another round to go through the tests.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1387 @@
+        /// behavior when used with [`Token`].
+        #[must_use]
+        pub(super) unsafe fn assume_exclusive_mutating_thread() -> Self {
+            thread_local! {
```
> WRT the ShortCounter `Token`-gating; the reason we can't or do not want to use an `AtomicU32` is simply because we want to have checked operations to handle overflow explicitly?

`AtomicU32` is slow, because it always emits atomic operations, necessitating the given cpu core to go to the global bus ;) by the way, `std::cell::Cell<ShortCounter>` can be used without runtime cost here, thanks to `ShortCounter: Copy`. But i wanted to dog-food `TokenCell` and to use `TokenCell`'s more concise api than `Cell`, which is made possible due to its additional constraints than `Cell`.

Also, note that `blocked_page_count` will be removed for upcoming opt.

ryoqun replied:
> As a summary, ...
> 
> As far as I can tell, ...

also, this is correct. thanks for wrapping up your understanding.

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -558,61 +578,78 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
-                        ))
-                        .unwrap();
-                    session_ending = false;
+                loop {
```
> We never exit this loop except for panic? I think we may have already talked about this in a previous PR, and you're just planning on error-handling in a follow-up.

This understanding is completely aligns with what i said previously. The proper scheduler management code is complicated, flavored to my own taste (expecting back-and-force review session), (i.e. yet another beast by itself). so, i wanted to focus on the logic with this pr.

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -558,61 +578,78 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
+                            recv(dummy_receiver(state_machine.has_unblocked_task())) -> dummy => {
+                                assert_matches!(dummy, Err(RecvError));
+
+                                if let Some(task) = state_machine.schedule_unblocked_task() {
```
done: 0a85082d5fea074ce3235ed467d22830c2fe5c09

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -558,61 +578,78 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
+                                let result_with_timings = result_with_timings.as_mut().unwrap();
+                                Self::accumulate_result_with_timings(result_with_timings, executed_task);
+                            },
+                            recv(dummy_receiver(state_machine.has_unblocked_task())) -> dummy => {
```
> The dummy receiver pattern seems really odd to me, though I see how it is working
>
> .........
>
> Anyway, my initial thought on the oddity of the pattern still stands. So not going to delete my comment, and still request your thoughts.

Hope I explained the justification clearly: 89b7773288187966419b9861d5d4fb0bd5981413. I think you can get used to it. this pattern is simplest and no redundant code and most performant. :)

> I'd actually much prefer we make this very clear since I don't think it's a very obvious thing:
> 
> ```rust
> let dummy_receiver = dummy_receiver(state_machine.has_unblocked_task());
> select!{
>     // ...
> }
> ```

This is done with comment. As I ranted in the comment, ideally, I want to move the `dummy_receiver()` invocation back into the `select!`...

> I know you had made a branch for a biased select - is the order here i.e. `finished -> dummy -> new` the biased order you want?

Yes, this is correct.



> It could be overkill for the current MNB load we have, but I wonder if you considered just running a hot loop here if we are actively scheduling; not sure if using try_recvs in hot loop would cause more contention on the channels, you're certainly more familiar with crossbeam internals.

As I explicit commented on in the commit 89b7773288187966419b9861d5d4fb0bd5981413. it's too early to take the busy loop path and i haven't fully grokked the perf implication of busy looping. Casual tests showed an improved latency indeed, though.

> But basically I am asking about something along the lines of:
> 
> ```
> ...
> ```

By the way, busy looping can be done with `select{,_biased}!` pretty easily like this:

```rust
loop ...
    select{_biased}! {
        recv(...) => ....,
        ...,
        default => continue, // there's special-cased the `default` selector in `select{,_biased}!`
    }
    ...
    if is_finished {
        break;
    }

}
```

ryoqun commented on Cargo.lock:
```diff
@@ -7462,7 +7462,9 @@ dependencies = [
 name = "solana-unified-scheduler-logic"
 version = "1.19.0"
 dependencies = [
+ "assert_matches",
```
btw, this will soon be not needed as par the progress at: https://github.com/rust-lang/rust/issues/82775

ryoqun commented:
> sans the big rename Page => UsageQueue and my take on comparison with prio-graph, i think i should have addressed all review comments so far. so, I'm requesting another review round. not sure you did go in depth for its algo code and my `unsafe`-related bla bla talks.. lol

I just finished the big renaming. still, i have yet to do the quoted homework of prio-graph. but, i think I've addressed all new review comments so far again. Always thanks for quite timely review!

willhickey commented:
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

ryoqun commented:
(@apfitzge although i migrated this pr over to https://github.com/anza-xyz/agave/pull/44, let's continue some chat in this pr for convenience..)

> (much easier with trimmed down PR! ❤️ ).

:heart: 

> ... As I'm thinking on this, I think this state-machine and the prio-graph I've used in the new scheduler for block-production are similar in many ways, although there are a few differences. Just gonna jot some thoughts here.
> 
> ## Similarities
> 
>   * Use FIFO fed order (in replay case this is record order, for production this is priority-order from a queue)
>   * Keep track of access-kind per account in "use"
>   * Pop unblocked txs
>
> ...
>
> ### Closing
> 
> Not sure anything I commented here has any actual value, just wanted to share my thoughts that these 2 approaches are remarkably similar

now that I've looked the prio-graph, I'm surprised by quite a bit of these similarities too. 


> ## Differences
> 
>  * PG essentially creates a linked-list of transaction dependencies to only **next** transaction.
>  * US keeps track of current access, and **all** blocked transactions/access type as an ordered list **per transaction**
>  * Blocked tx resolve time: PG resolves at insert time, US resolves at deschedule time.
>  * "use" differs since the US is used in a stateful manner, while PG is created on-demand

:100:  


> * PG has a mechanism for allowing user-defined ordering for "unblocked" txs

US will have this ability too for block production in the (hopefully not so long) future. Namaly, `UsageQueue::blocked_usages_from_tasks` will be `BTreeMap` via generics or something... That's needed for priority-based tx reordering obviously.
 
> 
> ### Example
> 
> Think visual examples always serve better than words here. Let's take a really simple example with 3 conflicting write transactions.
> 
> Prio-graph will effectively store a linked list, edges from 1 to 2 to 3.
> 
> ```mermaid
> graph LR;
> Tx1 --> Tx2 --> Tx3
> ```
> 
> US will initially store access of Tx1, and list w/ order (Tx2, Tx3). Only when Tx1 is descheduled will the conflict between 2 and 3 be "realized".
> 
> ```mermaid
> graph LR;
> Tx1 --> Tx2["Tx2(0, w)"] & Tx3["Tx3(1, w)"];
> ```

:100: (again quite good illustration of the differences)

Also note that delayed realization in US here is deliberate. so that late-arriving-yet-higher-prioritized tx should be able to cut the line as soon as the blocking-and-executing single tx is finished.

> these 2 approaches are ... different in their implmentation details due to different uses (stateful vs lazy).
> ... I actually made a branch in prio-graph to use this AccessList (Page?) pattern instead of directly tracking edges. 

fyi, I also considered the task-to-task edge approach. That said, i found 3 things are noteworthy from my perspective:

- PG's `pop` complexities is always smaller than US's `deschedule_task` because the former does need only look at conflicting addresses.
- on the other hand, as a pathological case, PG has quadratic (N*M) complexities in `insert_transaction` when write-locking many (N; N is up to 128) heavily read-locked address (M; unbounded). (this is one of the reasons; as US is stateful; this is especially important to avoid this..)
- Data ownership: `Arc` (or `Rc` soon) (US) vs Index based (PG). My gut tells the former is faster. yet I haven't tested it. lol

> Interestingly I see cases where the old approach does significantly better, but also cases where an approach similiar to yours does significantly better - interesting to see criterion regressions of 500% and also improvements of 50% in same bench run 😆

hehe, sounds a wild journey. this time, I've used `iai-callgrind` as primary opt work. As I shared my thinking for it previously elsewhere, it certainly had some quirks. but i found it generally savy to navigate towards well-rounded performance. with worst and normal cases.

> I am somewhat curious to try plugging in prio-graph to your benches to see how it compares for exactly the same benches you have. Do you have any benches on _just_ the scheduler logic?

i dunno you saw my prior comment (https://github.com/solana-labs/solana/pull/35286#issuecomment-1964383019) and you did actually tried this but please share if you find something interesting.

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+    }
+
+    #[test]
+    fn test_unblocked_task_related_counts() {
```
this seems like the same as the test above with one less task, and just checks some different counts on effectively same state? why not just check those counts in test above?

ryoqun replied:
wow. really thanks for looking into tests at that depth. I was just lazy.. now done: https://github.com/solana-labs/solana/pull/35286/commits/2cbf64a8cf4160ed05799133076798a7f0314cb4

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+    }
+
+    #[test]
+    fn test_all_blocking_redable_tasks_block_writable_task() {
```
typo:
```suggestion
    fn test_all_blocking_readable_tasks_block_writable_task() {
```

apfitzge commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+        assert_eq!(state_machine.active_task_count(), 1);
+        assert_eq!(state_machine.handled_task_count(), 2);
+        assert_eq!(state_machine.unblocked_task_queue_count(), 1);
+        // task3 is finally unblocked after all of readble tasks (task1 and task2) is finished.
```
typo:
```suggestion
        // task3 is finally unblocked after all of readable tasks (task1 and task2) is finished.
```

apfitzge reviewed (commented):
think i'm happy with the test cases, minus indexing ones (see comment).
However. I think we should get a 2nd opionion on all the unsafe code.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+            Self(UnsafeCell::new(value))
+        }
+
+        /// Returns a mutable reference with its lifetime bound to the mutable reference of the
+        /// given token.
+        ///
+        /// In this way, any additional reborrow can never happen at the same time across all
+        /// instances of [`TokenCell<V>`] conceptually owned by the instance of [`Token<V>`] (a
+        /// particular thread), unless previous borrow is released. After the release, the used
+        /// singleton token should be free to be reused for reborrows.
+        pub(super) fn borrow_mut<'t>(&self, _token: &'t mut Token<V>) -> &'t mut V {
+            unsafe { &mut *self.0.get() }
+        }
+    }
+
+    // Safety: Once after a (`Send`-able) `TokenCell` is transferred to a thread from other
+    // threads, access to `TokenCell` is assumed to be only from the single thread by proper use of
+    // Token. Thereby, implementing `Sync` can be thought as safe and doing so is needed for the
+    // particular implementation pattern in the unified scheduler (multi-threaded off-loading).
+    //
+    // In other words, TokenCell is technically still `!Sync`. But there should be no
+    // legalized usage which depends on real `Sync` to avoid undefined behaviors.
+    unsafe impl<V> Sync for TokenCell<V> {}
```
as par https://github.com/solana-labs/solana/pull/35286#pullrequestreview-1921305810:

> However. I think we should get a 2nd opionion on all the unsafe code.

@alessandrod This mod contains bunch of `unsafe`s and these highlighted two is the most important ones. I think I've extensively doc-ed it. Could you review whether my justification does make sense at all? I appointed you assuming you know Rust better than me from various memory mangling at vm. :) Happy to comment-in-source more to fill the inner workings of unified scheduler itself for general context. That said, feel free to defer to someone else. Thanks in advance.

alessandrod replied:
🫡

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+    }
+
+    #[test]
+    fn test_debug() {
```
> we don't rely on the exact format anywhere, right?

yes.

hehe, seems like i like coverage reports too much: https://github.com/solana-labs/solana/pull/35286/commits/3b9fb2cc5c65ce9aa8db459b51280c950c737e09; just removed.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1427 @@
+    fn test_conflicting_task_related_counts() {
+        let sanitized = simplest_transaction();
+        let address_loader = &mut create_address_loader(None);
+        let task1 = SchedulingStateMachine::create_task(sanitized.clone(), 101, address_loader);
+        let task2 = SchedulingStateMachine::create_task(sanitized.clone(), 102, address_loader);
+        let task3 = SchedulingStateMachine::create_task(sanitized.clone(), 103, address_loader);
+
+        let mut state_machine = unsafe {
+            SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+        };
+        assert_matches!(
+            state_machine
+                .schedule_task(task1.clone())
+                .map(|t| t.task_index()),
+            Some(101)
+        );
+        assert_matches!(state_machine.schedule_task(task2.clone()), None);
```
> Seeing these tests has made me somewhat concerned on the use of `index` within the state machine.
> The scheduler will schedule **not** based on index, but on the order in which transactions are fed in. It seems like the index is only tracked for sanity checks, but it feels like a layer violation

wow (again), quite good observation. Yeah, I agree. In retrospect, I added these sanity checks too prematurely, just because i can. I'll rethink on this later. for now, I just removed the assertions: d072efdd4d721613bd6ea89f96e3af060aa92888

(Lastly, I think you commented on wrong lines..)

ryoqun replied:
(this review thread is referenced by https://github.com/anza-xyz/agave/pull/3058#discussion_r1787839215)

ryoqun commented:
> think i'm happy with the test cases, minus indexing ones (see comment).

thanks for in-depth look into tests. I've addressed all.

> However. I think we should get a 2nd opionion on all the unsafe code.

kk. I've done this as well: https://github.com/solana-labs/solana/pull/35286#discussion_r1515573302

alessandrod commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+        /// instances of [`TokenCell<V>`] conceptually owned by the instance of [`Token<V>`] (a
+        /// particular thread), unless previous borrow is released. After the release, the used
+        /// singleton token should be free to be reused for reborrows.
+        pub(super) fn borrow_mut<'t>(&self, _token: &'t mut Token<V>) -> &'t mut V {
```
I understand the intent, but this is akin to a transmute and can be made
segfault trivially

```
struct T {
	v: Vec<u8>,
};

let mut token = unsafe { Token::<T>::assume_exclusive_mutating_thread() };
let c = {
	let cell = TokenCell::new(T { v: vec![42] });
	cell.borrow_mut(&mut token)
};
c.v.push(43);
```

ryoqun replied:
@alessandrod oh, quite nice catch.. hope this fixes segfault and no more ub?: 001b10e17d6444ba296c27d25191408f614bdf85

seems i'm too short-sighted only with aliasing-based ub.. ;)

ryoqun replied:
uggg, actually, i need `with_borrow_mut()`.. I'll soon push this.

ryoqun replied:
> uggg, actually, i need `with_borrow_mut()`.. I'll soon push this.

done: 02567b18b28e97bcb0d3319b190467f58b2b461d

alessandrod reviewed (commented):
Finished my first pass at this. I really like the general structure and direction! Being completely new to this new scheduler code I've struggled with some of the terminology and I've left some nits about that. Also see the TokenCell issue.

ryoqun commented:
> I really like the general structure and direction!

:heart:

> Being completely new to this new scheduler code I've struggled with some of the terminology and I've left some nits about that.

these feedback is quite appreciated. I'll address in turn to fill the missing contexts.

> Also see the TokenCell issue.

as this is the most important review comment, I've address it firstly, and requested review for it...

ryoqun commented:
gg

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
 
-pub struct Task {
+/// Internal utilities. Namely this contains [`ShortCounter`] and [`TokenCell`].
+mod utils {
```
> You don't like files? :P

nope. I don't like large diffs, especially if that code shuffling hides bunch of review threads. :)



> was pretty weird to see. Especially since the file starts with the utils, which
> arguably are not the thing I care the most about if I'm checking out the
> scheduler logic.

voice heard; maybe, will do this as a follow up pr.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+/// `solana-unified-scheduler-pool`.
+pub struct SchedulingStateMachine {
+    unblocked_task_queue: VecDeque<Task>,
+    active_task_count: ShortCounter,
```
hehe, i agree tastelessness of `active` here. hmm, `current_task_count` is a bit ambiguous. how about `in_progress_task_count`? (or `not_(yet)_handled_task_count`; don't like this much because of implication of being a counter of tasks which has been resolved not to process).

alessandrod replied:
I find in_progress as confusing, since it implies that something is progressing while blocked tasks technically aren't progressing. @apfitzge as the native speaker, wdyt would be the best term here?

ryoqun replied:
how about `pending_task_count`, or `managed_task_count`, then?

ryoqun replied:
... or `runnable_task_count`?

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+//!
+//! The most important type is [`SchedulingStateMachine`]. It takes new tasks (= transactions) and
+//! may return back them if runnable via
+//! [`::schedule_task()`](SchedulingStateMachine::schedule_task) while maintaining the account
```
i know this is my peculiarities. but, I like shorter prose in html.

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    /// Returns `Some(task)` if it's immediately scheduled. Otherwise, returns `None`,
+    /// indicating the task is blocked currently.
+    ///
+    /// Note that this function's type signature is intentionally redundant to take the ownership
```
yeah. just simplified: f97ecc6c8163c6795e89fedbca11f2a7c0f431bf

maybe, I was too dragged down by the context when i added this para. ref: https://github.com/solana-labs/solana/pull/35286#discussion_r1499924386

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+        })
+    }
+
+    pub fn deschedule_task(&mut self, task: &Task) {
```
good question: 250edde6dde10899ffb7c05d1b143fca4b0f1705

> If not, the caller can re-schedule the same task, which seems wrong API wise

Even if this fn consumes the task, this would be possible because `Task` is `Clone`... That said, I'm planning to make it `!Clone`.

alessandrod replied:
From the referenced commit 

>     /// Note that this function intentionally doesn't take ownership of the task to avoid dropping
>    /// tasks inside `SchedulingStateMachine` to provide an offloading-based optimization
 >   /// opportunity for callers.

Could you say more about what optimization you have in mind? Documenting invariants is good, but type system > docs, so if we can, we should enforce this at the type level IMO.

(And agreed that tasks shouldn't be Clone)

ryoqun replied:
> Could you say more about what optimization you have in mind?

well, i thought i was explicit. avoiding dropping and offloading that to other background thread is an optimization. as a general context, the scheduler thread itself is the weakest point (bottleneck) of my unified scheduler thing. its performance is very susceptible to any overhead...

ryoqun replied:
btw, the optimization will follow soon because I'm now in to mood of missing 1.19 (or 2.0?) branching.

> (And agreed that tasks shouldn't be Clone)

btw, this will be fixed together.

alessandrod replied:
> well, i thought i was explicit. avoiding dropping and offloading that to other background thread is an optimization. 

It's not explicit because from the existing code it's not obvious what the optimization would even be. Currently the thread calling deschedule_task() is in fact dropping the Task (presumably - because to further complicate things, Task is an Arc, so who knows who's holding refs). This btw is why personally I'm not a fan of landing big changes piecemeal - it's easier to review code in its proper context than in isolation and then "things will happen in the future".

Anyway, whatever is happening to the task, surely this is something that can be expressed through the type system? Eg if after deschedule_task(task) the task can't be scheduled anymore, deschedule_task could take by value and return a marker Completed(task) type which makes it impossible to access the inner value. Then Completed(task) can still be dropped in another thread.

ryoqun replied:
> It's not explicit because from the existing code it's not obvious what the optimization would even be. Currently the thread calling deschedule_task() is in fact dropping the Task (presumably - because to further complicate things, Task is an Arc, so who knows who's holding refs). This btw is why personally I'm not a fan of landing big changes piecemeal - it's easier to review code in its proper context than in isolation and then "things will happen in the future".

yeah. sorry for extra trouble... as you noticed, it's currently dropping in-thread indeed. and that's because of piecemeal approach.. i know these back-and-fort aren't fun. however, i still like this approach as personal preference.

> Anyway, whatever is happening to the task, surely this is something that can be expressed through the type system? ...

yep, i agree. thanks for suggestion. I'll do once after https://github.com/anza-xyz/agave/pull/129#issuecomment-2016423148 is settled on not trashing this whole involved approch in this pr...

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+            //
+            // Note that large cap should be accompanied with proper scheduler cleaning after use,
+            // which should be handled by higher layers (i.e. scheduler pool).
+            blocked_usages_from_tasks: VecDeque::with_capacity(128),
```
yep. in the (near) future. :)

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+//! [`::schedule_task()`](SchedulingStateMachine::schedule_task) while maintaining the account
+//! readonly/writable lock rules. Those returned runnable tasks are guaranteed to be safe to
+//! execute in parallel. Lastly, `SchedulingStateMachine` should be notified about the completion
+//! of the exeuction via [`::deschedule_task()`](SchedulingStateMachine::deschedule_task), so that
```
kitten dies. :crying_cat_face: well, it's really bad condition, i haven't thought about  in depth. it will easily lead to dead locks.

well, i wonder this edge case should be called out in this summary doc.. fyi, I explicit mentioned this case in the `deschedule_task()` doc comment: 250edde6dde10899ffb7c05d1b143fca4b0f1705

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    }
+
+    #[must_use]
+    pub fn schedule_unblocked_task(&mut self) -> Option<Task> {
```
unlike those verbose names you pointed here (https://github.com/solana-labs/solana/pull/35286#discussion_r1527128326), this is `pub` function, so I rather keep internal details hidden. So, I intentionally chose a generic name here. 

in other words, as long as scheduler pool sees `.has_unblocked_task()`, it just calls `schedule_unblocked_task()` without bothering which kind of those unblocked task are and in which order.

alessandrod replied:
I'm not sure why we would optimize for number of characters vs clarity :P To me, schedule_unblocked_task reads "schedule _the_ task which is now unblocked", not "schedule _any_ unblocked task".

I'd prefer explicit name, but at the very minimum docs should clarify.

ryoqun replied:
> To me, schedule_unblocked_task reads "schedule _the_ task which is now unblocked", not "schedule _any_ unblocked task".

oh, thx for feedback. in that case, that's not my intention: 3f233501760

> I'd prefer explicit name, but at the very minimum docs should clarify.

agree. :)

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+                }
+                LockResult::Err(()) => {
+                    blocked_usage_count.increment_self();
+                    let usage_from_task = (attempt.requested_usage, task.clone());
```
oh, I like `.clone()` being actually `Arc::clone()` behind scene to hide this (minor) implementation detail.

alessandrod replied:
Agree to disagree on this one I guess - when I see a clone() I think "this might be expensive and there might be an optimization possible in removing the clone", while if I see Arc::clone() I know it's not expensive.

ryoqun replied:
> if I see Arc::clone() I know it's not expensive.

hehe, fun fact is that `Arc::clone()` is super slow in this unified scheduler code.

alessandrod replied:
"super slow" is a bold claim that requires extraordinary evidence 😉

ryoqun replied:
hehe, true.

how about this? i think 1.7x perf increase should be a big deal: https://github.com/solana-labs/solana/pull/35286#issuecomment-1964383019 just with this code changes https://github.com/ryoqun/solana/pull/17#discussion_r1536575150 (with `unsafe`, unfortunately)

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    #[must_use]
+    fn unlock_usage_queue(
+        usage_queue: &mut UsageQueueInner,
+        attempt: &LockAttempt,
```
thanks for nice suggestion: 34f56506f49447daeba32a6109cd81ad3c6d40b1

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    }
+
+    #[must_use]
+    fn try_lock_for_task(&mut self, task: Task) -> Option<Task> {
```
or try_lock_task_resources or try_lock_task_usages?

ryoqun replied:
see also my comment: https://github.com/solana-labs/solana/pull/35286#discussion_r1529863664

alessandrod replied:
try_lock_task_resources is by far my favorite, but kinda strange if you want to keep UsageQueue. Until it's UsageQueue, for consistency I think try_lock_usages would be better

ryoqun replied:
done (actually went with `*_usage_queues()`): 8c0c6bddba52be5a81a1fad4efd9f878559f1374

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+
+            while let Some((requested_usage, task_with_unblocked_queue)) = unblocked_task_from_queue
+            {
+                if let Some(task) = task_with_unblocked_queue.try_unblock(&mut self.count_token) {
```
another nice question. hope this helps your understanding: ee5c8b69be0698d70da53cce25c5761706dfcce4

alessandrod replied:
Can you explain how this works? usage_queue.unlock() pops the task from the usage queue. task.try_unblock() consumes self and returns None. How does the task reappear?

alessandrod replied:
I see. This is why I dislike hiding Arcs with type aliases :P

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+                LockResult::Err(())
+            };
+            match lock_result {
+                LockResult::Ok(Usage::Unused) => unreachable!(),
```
oh, cool. nice observation: aea6a24f4972f2bfb45d48ff6dd9e75f1ce26c37

yeah, i considered that loong time ago. seems some cleaning up made `Option<Usage>` break tie with `Usage::Unused`. :)

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    }
+
+    #[must_use]
+    fn pop_unblocked_usage_from_task(&mut self) -> Option<UsageFromTask> {
```
thanks for suggestion. I added assertions: 04b65a4de109be9a2f74357794488a98696e9422

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    }
+
+    #[test]
+    fn test_scheduling_state_machine_reinitialization() {
```
done: 0056cfc2cc1734057e2f20c7dc4923d9c3b2bd16

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+/// [`Task`]'s per-address attempt to use a [usage_queue](UsageQueue) with [certain kind of
+/// request](RequestedUsage).
+#[derive(Debug)]
+struct LockAttempt {
```
yep, `LockAttempt` was misnomer...

> , as in: attempt to lock a few times.

haha, this feeling is right. previously, this did actually attempt to lock a few times. not it's gone. so, name is only left behind....

how about `LockContext`?: cff8ad8a4d29f00c46918cb094a465c90907c0a0

It's a bit odd to me for `LockRequest` to hold some contextual data (`.usage_queue`), which is preloaded in the uniform way, regardless of specifics of individual tasks. say, `HttpRequest` is holding some database record.

I think `LockContext` is still consistent with `LockResult`, `RequestedUsage`.

alessandrod replied:
Not a fan of Context in general as a name, but that's just a matter of personal taste, it's clearer than LockAttempt so lgtm

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    /// separation of concern.
+    pub fn create_task(
+        transaction: SanitizedTransaction,
+        index: usize,
```
added docs: 0a9421aa6083628fc87c678b12a883508a0086e2

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+    ) -> Option<(RequestedUsage, Task)> {
+        let mut is_unused_now = false;
+        match &mut usage_queue.current_usage {
+            Usage::Readonly(ref mut count) => match attempt.requested_usage {
```
well, it seems that i like those explicit `match` arms even if they're nested.

alessandrod replied:
I thought you liked conciseness!? 😉

ryoqun replied:
lol. seems my peculiar case-by-case aesthetics can't be deciphered.

i like conciseness. but i like consistency more. dunno you can get it. but, in this case, `Usage::Readonly` and `Usage::Writable` has equal logical significance to me. so, it must be treated same. i have some personal taste to write

```rust
# when the two branches have same significance.
if ... {
   do_something1();
} else {
   do_something2();
}
```

or 
```rust
# when the two branches don't have same significance
if ... {
    do_something1();
    return;
}
do_something2();
```

plus, the nested syntax can realize consistency with `try_lock`. Also, i like exhaustiveness of match for future maintainability of new variants.

That said i intentionally breaking consistency of ordering each arms in `try_lock` and `unlock`. that reveals my third criteria: normal case first and state transition flow. maybe i shouldn't waste my worktime anymore. lol

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+#[derive(Debug)]
+struct UsageQueueInner {
+    current_usage: Usage,
+    blocked_usages_from_tasks: VecDeque<UsageFromTask>,
```
> In general I find UsageQueue* a bit confusing, since it puts the emphasis on a
> secondary thing. To me the main thing is that the queue holds tasks that are
> waiting. Then which access mode those tasks require is secondary.

hmm, i understand your reasoning. however, seems that there's some disagreement here. Maybe, it's just that I'm very reluctant to say something as `task`, which is actually a tuple of `(Task, <readonly or writable>)`. And treating access mode (or access request in my cognition) as primary thing isn't so bothering to me.

> Alternatively the more I read the code, the more I think UsageQueue* could be
called Resource

hehe, i and andrew actually had tangent discussion. Previously, `UsageQueue` was called `Page`.: https://github.com/solana-labs/solana/pull/35286#discussion_r1501284792

I think `Resource` is too generic in similar vein. i want to express how things are actually used or implemented with names.

> Then you could have lock_task_resources etc.

as i wrote here https://github.com/solana-labs/solana/pull/35286#discussion_r1528758840, `{lock/unlock}_task_usages` isn't super confusing to me. admittedly, I'm in search of more good names to replace `usage`, though.

alessandrod replied:
Wish I could go back and vote for LockQueue :P 

Personally I really dislike usage, and lock/unlock_task_usages doesn't sound like proper english to me. That said it sounds like names aren't set in stone yet, so we can get the logic in and continue the naming conversation at a later stage I guess.

ryoqun replied:
how about replacing `UsageQueue` with `AddressLock` and rename things all?

ryoqun commented on unified-scheduler-logic/src/lib.rs:
```diff
@@ -1,20 +1,1299 @@
+
+        for attempt in task.lock_attempts() {
+            let usage_queue = attempt.usage_queue_mut(&mut self.usage_queue_token);
+            let lock_result = if usage_queue.has_no_blocked_usage() {
```
see my comment: https://github.com/solana-labs/solana/pull/35286#discussion_r1529863664

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -558,61 +588,97 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
+                                let result_with_timings = result_with_timings.as_mut().unwrap();
+                                Self::accumulate_result_with_timings(result_with_timings, executed_task);
+                            },
+                            recv(dummy_unblocked_task_receiver) -> dummy => {
+                                assert_matches!(dummy, Err(RecvError));
+
+                                let task = state_machine.schedule_unblocked_task().expect("unblocked task");
+                                runnable_task_sender.send_payload(task).unwrap();
+                            },
```
here
